refactor strongTypePtr

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32959 jit pickling rref
* **#33590 refactor strongTypePtr**

Differential Revision: [D20015521](https://our.internmc.facebook.com/intern/diff/D20015521/)

zhaojuanmao committed:
refactor strongTypePtr



Differential Revision: [D20015521](https://our.internmc.facebook.com/intern/diff/D20015521/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "refactor strongTypePtr"


Differential Revision: [D20015521](https://our.internmc.facebook.com/intern/diff/D20015521/)

[ghstack-poisoned]

suo commented on test/cpp/jit/CMakeLists.txt:
```diff
@@ -27,6 +27,8 @@ elseif (USE_ROCM)
 
   target_link_libraries(test_jit PRIVATE caffe2_gpu)
 
+  target_link_libraries(test_jit PRIVATE torch_cpu)
```
What is the purpose of this?

zhaojuanmao replied:
it is trying to resolve link error, but it seems not to resolve the failures in CI

suo commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -304,6 +307,17 @@ std::vector<std::pair<IValue, IValue>> iterationOrder(const c10::Dict<IValue, IV
+  type_ = type;
+  TORCH_INTERNAL_ASSERT(type_);
+  if (type_->cast<ClassType>()) {
+    TORCH_INTERNAL_ASSERT(cu_);
```
Can we add a more detailed error message in this case? Also, one useful check might be to assert that the class type's owning compilation unit is the same as the `cu` we passed in here.

zhaojuanmao replied:
will do

suo commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -656,13 +656,10 @@ struct CAFFE2_API WeakIValue final {
 struct StrongTypePtr {
```
Can you update the comment for this struct?

zhaojuanmao replied:
will do

zhaojuanmao committed:
Update on "refactor strongTypePtr"


Differential Revision: [D20015521](https://our.internmc.facebook.com/intern/diff/D20015521/)

[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot retest this please
